[Fiber] setState

acdlite opened the pull request:
**`workInProgress`** (yay puns)

Implements `setState` for Fiber. Updates are scheduled by setting a work priority on the fiber and bubbling it to the root. Because the instance does not know which tree is current at any given time, the update is scheduled on both fiber alternates.

Need to add more unit tests to cover edge cases.

A few questions I have:

- How to deal with the circular dependency between ReactFiberScheduler and ReactFiberBeginWork. I'm using a closure to lazily access the scheduler but seems like there may be a better way.
- How to queue update callbacks. The only way I could think to do it was to add another field to Fiber, so I've punted on this for now.
- How to access the fiber from the updater. I decided to use a private field on the instance. Could also create a new updater for each instance and close over its fiber, but that seems wasteful.
- Should I put `setState` related tests in their own module? I think I should but not sure.

(I discussed with @sebmarkbage before working on this PR.)

Based on https://github.com/facebook/react/pull/7448. Compare view: https://github.com/sebmarkbage/react/compare/fiberstarvation...acdlite:fibersetstate

acdlite commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -49,41 +52,125 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
+    enqueueSetState(instance, partialState) {
+      const fiber = instance._fiber;
+
+      const prevState = fiber.pendingState || fiber.memoizedState;
```
Pretty sure using `fiber.memoizedState` here is a bug because we don't know if `fiber` is current. But, I don't know how to get the current state. It would be nice if we could rely on `instance.state` but that won't work in the case of an aborted update.

acdlite replied:
Will have the same problem with getting the current props for the updater form of  `setState((state, props) => newState)`.

sebmarkbage replied:
Yea, that's why this will need to become a queue of state transitions for those cases.

You may also need to be able to schedule different priority level state transitions on the *same* component. In that case the queue can be reordered. We can wait on that though. Not sure it is worth the trouble.

Can `pendingState` just use the partial state and that way you can do the merge at the time you apply it?

sebmarkbage commented:
We'll figure something out with the dependency later. One technique is to just put everything in the same file and then try to break it apart again into reasonable pieces.

I figured that the simple `pendingState` could an object, and then if you use `setState((state, props) => ...);` or the callback model or it would turn into an array or linked list.

To access the fiber from the updater, you can use a field. That's what we do now. However, you could switch the implementation to use `ReactInstanceMap` which the current implementation does so we can share that code. Just move it to: `src/renderers/shared/shared/` to indicate that we would like to share it between both implementations.

sebmarkbage commented:
As for the unit tests, we can keep them in Incremental for now. In the near future we'll just run this against the existing unit tests for `setState`. We'll evolve the tests you're writing now into tests that specifically tests the new incremental effects of setState.

acdlite commented:
Ah keeping a queue of pending states makes sense. A linked list sounds like the way to go, but how to disambiguate between a state object and a linked list? `$$typeof`?

sebmarkbage commented:
Always wrapping it in a linked list node might be fine. Should be short lived and not a very hot path.

> On Jul 24, 2016, at 4:44 PM, Andrew Clark <notifications@github.com> wrote:
> 
> Ah keeping a queue of pending states makes sense. A linked list sounds like the way to go, but how to disambiguate between a state object and a linked list? $$typeof?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

acdlite commented:
Changed the type of `pendingState` to be a queue.

Regarding typing, I'm guessing eventually the type of the props and state will be inferred from the element? For now I've used `any` because that's what props uses, but I know that's bad.

Also not sure about naming. `pendingState` is confusing because it's a queue of partial states, whereas `pendingProps` is a single props object.

acdlite commented:
Maybe rename `pendingState` to `stateQueue`?

sebmarkbage commented:
Sure. Wouldn't worry too much about it yet until all the pieces are there and have proven themselves.

It's quite possible that props will become a queue too if we want to solve the problem of props not being consistent when an event happens.

> On Jul 24, 2016, at 11:04 PM, Andrew Clark <notifications@github.com> wrote:
> 
> Maybe rename pendingState to stateQueue?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

acdlite commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -49,41 +57,142 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
+      // is current.
+      scheduleUpdate(fiber, pendingState, LowPriority);
+      if (fiber.alternate) {
+        scheduleUpdate(fiber.alternate, pendingState, LowPriority);
```
Rather than bubble up the priority both trees, could we bubble once and assign to the alternate at each level? I think so...

sebmarkbage replied:
Yea, that's the only correct way to do it, because `.return` is not guaranteed to point to two different nodes even for two children.

These are not two entirely parallel trees. They can overlap. Two parent fibers can point to the same exact child fiber when the child gets reused.

acdlite replied:
Hmm, I'm confused. I see your point about how the trees aren't parallel, but given that, isn't bubbling up both trees the only way? The first part of your comment seems to contradict the second.

sebmarkbage replied:
The parent path is guaranteed to be the same conceptual component path. However, you can't rely on any particular node being the current one along the path.

So this is safe:

```js
while (node) {
  addState(node);
  addState(node.alternate);
  node = node.return;
}
```

But this is not safe, because it is not guaranteed to cover every node:

```js
node = start;
while (node) {
  addState(node);
  node = node.return;
}
node = start.alternate;
while (node) {
  addState(node);
  node = node.return;
}
```

An example when this happens:

```js
// Assume the cycles are resolved
a1 = { child: b1, return: null, alternate: a2 };
a2 = { child: b2, return: null, alternate: a1 };
b1 = { child: c1, return: a1, alternate: b2 };
b2 = { child: c1, return: a2, alternate: b1 };
c1 = { child: null, return: b1, alternate: c2 };
c2 = { child: null, return: b1, alternate: c1 }; // Note that the return is the same b1.
```

Since two versions of a fiber can point to the same child, it is not possible for a parent pointer alone to point to every parent since there can be multiple.

Conceptually this pointer is used for two purposes. A pointer to the parent "Instance" which I've merged into the Fiber. See the type definition.

However, it is also used as a temporary pointer for knowing which fiber to step back through after processing.

If I wasn't so unnecessarily clever about allocations the structure would like this:

```js
type Instance = { parent: ?Instance };
type Fiber = { inst: Instance, return: ?Fiber };
```

and you would just walk the parent pointer. However, I'm being clever to see how far we can get but it adds some complexity.

acdlite replied:
Ooooh, okay. I think I get it now! Thanks for the detailed explanation (and for being patient as I try to keep up, haha).

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberStateQueue.js:
```diff
@@ -37,16 +46,17 @@ exports.addToQueue = function(queue : StateQueue, partialState : mixed): StateQu
-      queue.partialState(state, props) :
-      queue.partialState;
+    const partialState = typeof node.partialState === 'function' ?
+      node.partialState(state, props) :
```
This will pass the `node` to user code, leaking this info. Pick off the function before invoking it so that `this` will be undefined (or global depending on strict mode).

```js
const stateUpdater = node.partialState;
const partialState = stateUpdater(state, props);
```

acdlite replied:
Good catch

acdlite commented:
A few things left to do.

We need a separate field for the priority of a fiber's pending props and state, distinct from the existing field `pendingWorkPriority`, which represents the highest priority of the entire subtree. Otherwise, there's no way to schedule work deep in the tree without overriding the priority of every ancestor. I got this working locally but I'm holding off until @sebmarkbage fixes some priority-related bugs on his branch.

We also need more tests, specifically around preempted updates. In order to do this, we need to implement high priority updates. @sebmarkbage's idea is an API like `ReactNoop.performHighPriWork(fn)` where all updates called within that scope have high priority by default.

acdlite commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -132,6 +134,29 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>, getSchedu
+      if (fiber.callbackList) {
+        // If this fiber was preempted and already has callbacks queued up,
+        // we need to make sure they are part of the new update queue
+        updateQueue = concatQueues(fiber.callbackList, updateQueue);
```
Oops, this isn't right because this also adds the state updates the beginning of the queue! I believe this is unobservable unless there are side-effects inside state updaters (which there shouldn't be), but it's wasteful regardless.

Should instead combine the callbacks into a single callback, add to an empty node, and prepend it to the update queue.

acdlite commented:
@sebmarkbage Rebased

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -266,7 +266,8 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
         // pending props from the work in progress. Needs restructuring.
-        return findNextUnitOfWorkAtPriority(workInProgress, priorityLevel);
+        var work = findNextUnitOfWorkAtPriority(workInProgress, priorityLevel);
+        return work;
```
This change was reverted but still in your commit history due to the rebase not being able to ignore.

acdlite replied:
Oops. I think it's fixed now?

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberScheduler.js:
```diff
@@ -126,6 +136,10 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
       workInProgress.pendingProps = null;
+      workInProgress.updateQueue = null;
+      if (current) {
+        current.updateQueue = null;
```
This looks scary. This whole work in progress tree can still be aborted so it is not safe to reset this yet. Wouldn't that drop work if that happens?

acdlite replied:
Is the fix as simple as not resetting the update on current here? I think so...

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberCompleteWork.js:
```diff
@@ -132,6 +130,17 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
+        // called during commit phase.
+        workInProgress.callbackList = workInProgress.updateQueue;
+        if (current) {
+          current.callbackList = workInProgress.callbackList;
```
Same issue here. We should be able to drop the whole work in progress tree without any side-effects being retained.

acdlite replied:
Hmm I think I see what you mean. So the only place we should ever reset the queue / callback list is after the changes are committed?

acdlite replied:
Ah alright. I'll need to rethink what to do here then: https://github.com/acdlite/react/blob/04c9a399e871d5ffe7323f11132b51185bc32e6e/src/renderers/shared/fiber/ReactFiberBeginWork.js#L158-L174

Does the overall strategy of callback list seem sound or should I go back to the drawing board?

acdlite commented:
@sebmarkbage Pushed some changes that I believe address your comments. I think I understand the relationship between the current tree and work in progress tree better now.

acdlite commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -79,31 +92,143 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
+      if (preemptedFiber = fiber.callbackList || (fiber.alternate && fiber.alternate.callbackList)) {
+        // If the workInProgress was preempted and already has callbacks queued
+        // up, we need to make sure they are part of the new update queue.
+        const { callbackList } = preemptedFiber;
```
Actually... I don't think any of this is necessary at all, right? If an update is preempted the callbacks won't be lost because they're still on the current tree's update queue?

acdlite replied:
Difficult for me to reason about this since for now it's not yet possible to test :D

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberCommitWork.js:
```diff
@@ -31,6 +32,10 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
       case ClassComponent: {
+        if (finishedWork.callbackList) {
+          callCallbacks(finishedWork.callbackList);
+          finishedWork.callbackList = null;
```
I tend to prefer extract such lists off the object and reset it to null before calling it. Makes it easier to reason about things that happen inside of the callbacks, such as new callbacks being scheduled onto this list or errors being thrown.

acdlite replied:
Good call

acdlite commented:
@sebmarkbage I rebased against your fiberstarvation branch. I think you're right to worry about sharing the update queue between both trees — after rebasing, there was a bug where updates weren't being cleared from the current tree. I solved by [clearing the queue during the commit phase](https://github.com/facebook/react/pull/7344/commits/a6dc55f87228d04ac3b476dbb8dd296eaf288236#diff-7002876230f7effe9aa22e292f6a52c6R35) but I'm unsure if it's safe to do so. Added a test to ensure that `setState` works inside a callback.

sebmarkbage commented:
I rebased this on top of my stuff for you. However, now it is failing tests. Not sure why yet.

https://github.com/sebmarkbage/react/tree/fibersetstate

sebmarkbage commented on src/renderers/shared/fiber/ReactFiberBeginWork.js:
```diff
@@ -201,113 +308,82 @@ module.exports = function<T, P, I, C>(config : HostConfig<T, P, I, C>) {
+    if (workInProgress.pendingProps === null || (
+      workInProgress.memoizedProps !== null &&
+      workInProgress.pendingProps === workInProgress.memoizedProps &&
+      workInProgress.updateQueue === null
```
I screwed this up in the rebase. If `pendingProps === null` the updateQueue also needs to be `=== null`. I'll fix it.

sebmarkbage commented:
We know that there might still be issues with some minor details of this and we'll need more tests, but this is a great start. I'll accept it pending that the base PR https://github.com/facebook/react/pull/7636 lands.

sebmarkbage commented:
Ninja merging this. Will fix travis if errors are caused. #workedonmymachine
